### ISSUE TITLE
Decimal compare string parameter can't be pushed down when using plan cache

### ISSUE THREAD

**Issue by time-and-fate** [pingcap/tidb#28064]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
enable plan cache
```sql
CREATE TABLE `t` (
  `a` decimal(10,0) DEFAULT NULL,
  `b` decimal(10,0) DEFAULT NULL,
  `c` decimal(10,0) DEFAULT NULL,
  `d` decimal(10,0) DEFAULT NULL,
  KEY `iabc` (`a`,`b`,`c`)
);
set @a='123', @b='234', @c='345';
prepare stmt1 from 'select * from t use index (iabc) where a = ? and b = ? and c = ?';
execute stmt1 using @a, @b, @c;
```

### 2. What did you expect to see? (Required)
```
+-------------------------------+---------+-----------+------------------------------+-----------------------------------------------------------------+
| id                            | estRows | task      | access object                | operator info                                                   |
+-------------------------------+---------+-----------+------------------------------+-----------------------------------------------------------------+
| IndexLookUp_7                 | 0.00    | root      |                              |                                                                 |
| ├─IndexRangeScan_5(Build)     | 0.00    | cop[tikv] | table:t, index:iabc(a, b, c) | range:[123 234 345,123 234 345], keep order:false, stats:pseudo |
| └─TableRowIDScan_6(Probe)     | 0.00    | cop[tikv] | table:t                      | keep order:false, stats:pseudo                                  |
+-------------------------------+---------+-----------+------------------------------+-----------------------------------------------------------------+
```

### 3. What did you see instead (Required)
```
+---------------------------------+----------+---------+-----------+------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------+--------+------+
| id                              | estRows  | actRows | task      | access object                | execution info                                                                                                                                                  | operator info                                           | memory | disk |
+---------------------------------+----------+---------+-----------+------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------+--------+------+
| Selection_5                     | 8000.00  | 0       | root      |                              | time:1.37ms, loops:1                                                                                                                                            | eq(test.t.a, 123), eq(test.t.b, 234), eq(test.t.c, 345) | N/A    | N/A  |
| └─IndexLookUp_8                 | 10000.00 | 0       | root      |                              | time:1.36ms, loops:1,  table_task: {total_time: 5.03ms, num: 0, concurrency: 5}                                                                                 |                                                         | N/A    | N/A  |
|   ├─IndexFullScan_6(Build)      | 10000.00 | 0       | cop[tikv] | table:t, index:iabc(a, b, c) | time:682.4µs, loops:1, cop_task: {num: 1, max: 397µs, proc_keys: 0, rpc_num: 1, rpc_time: 333.6µs, copr_cache_hit_ratio: 0.00}, tikv_task:{time:0s, loops:0}    | keep order:false, stats:pseudo                          | N/A    | N/A  |
|   └─TableRowIDScan_7(Probe)     | 10000.00 | 0       | cop[tikv] | table:t                      |                                                                                                                                                                 | keep order:false, stats:pseudo                          | N/A    | N/A  |
+---------------------------------+----------+---------+-----------+------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------+--------+------+
```

### 4. What is your TiDB version? (Required)

master

**Comment by time-and-fate** [comment]
This bug is triggered by several steps:
1. In expression rewriter, we set default `FieldType` for the param marker `Expression` in `ParamMarkerExpression()`. Here we set its `RetType`'s `Flen` and `Decimal` to `UnspecifiedLength`, which is `-1` because the param is a string.
2. Still in expression rewriter, then we build `Expression` for the `BinaryOperationExpr`, which is in `binaryOpToExpression() => constructBinaryOpFunction() => newFunction()`. Because this is a decimal column comparing string constant, we need to wrap the string constant with a "cast as decimal". Here we directly set the `Flen` and `Decimal` of `RetType` of the cast function with the `Flen` and `Decimal` from the string constant. 
This is at https://github.com/pingcap/tidb/blob/0f51627cd1386cd87cc642633fca4abdccb8cb58/expression/builtin_cast.go#L1898
3. As we know, we will fold constant when building a new scalar function. So here this cast function becomes a `Constant` with `DeferredExpr` set to the cast function and `RetType`'s `Flen` and `Decimal` set to -1.
3. In `(*DataSource).PredicatePushDown`, we try to push down the constant. To check if it can be pushed down, we will first eval it and then call `encodeDatum()` in `(PbConverter).conOrCorColToPBExpr`.
When eval it, because it has `DeferredExpr`, there is an extra step, which is to convert the result to `RetType` of the `Constant`.
This is at https://github.com/pingcap/tidb/blob/0f51627cd1386cd87cc642633fca4abdccb8cb58/expression/constant.go#L200
In it we directly set `RetType.Flen`(int) and `RetType.Decimal`(int) to `Datum.length`(uint32) and `Datum.decimal`(uint16). So the result datum's `length` becomes 4294967295 and `decimal` becomes 65535.
Because this is a `MyDecimal` with invalid `digitsInt` and `digitsFrac`, it returns an error in `encodeDatum()`, and pushing down fails.
(In fact, the invalid `digitsFrac` is automatically changed to `mysql.MaxDecimalScale` in `EncodeDecimal()`, which is valid. But `digitsInt` is still invalid, so there is still an error)
